src/Compilers/Core/MSBuildTask/DotnetHost.cs

```diff
@@ -0,0 +1,55 @@
+﻿using System;
+
+namespace Microsoft.CodeAnalysis.BuildTasks
+{
+    class DotnetHost
+    {
+        public string ToolName { get; }
+
+        public string PathToTool { get; }
+
+        public string CommandLineArgs { get; }
+
+        public DotnetHost(string toolName, string pathToTool, string commandLineArgs)
+        {
+            ToolName = toolName;
+            PathToTool = pathToTool;
+            CommandLineArgs = commandLineArgs;
+        }
+
+        public DotnetHost(string toolNameWithoutExtension, string commandLineArgs)
+        {
+            string pathToTool;
+            string toolName;
+            if (IsCliHost(out string pathToDotnet))
+            {
+                toolName = $"{toolNameWithoutExtension}.dll";
+                pathToTool = Utilities.GenerateFullPathToTool(toolName);
+                if (pathToTool != null)
+                {
+                    RewriteToCliInvocation(ref pathToTool, ref commandLineArgs, pathToDotnet);
+                }
+            }
+            else
+            {
+                // Desktop executes tool directly
+                toolName = $"{toolNameWithoutExtension}.exe";
+                pathToTool = Utilities.GenerateFullPathToTool(toolName);
+            }
+            PathToTool = pathToTool;
+            CommandLineArgs = commandLineArgs;
+        }
+
+        private static void RewriteToCliInvocation(ref string pathToTool, ref string commandLineArgs, string pathToDotnet)
+        {
+            commandLineArgs = $"\"{pathToTool}\" {commandLineArgs}";
+            pathToTool = pathToDotnet;
+        }
+
+        private static bool IsCliHost(out string pathToDotnet)
+        {
+            pathToDotnet = Environment.GetEnvironmentVariable("DOTNET_HOST_PATH");
```

Comment:
Hmm. I worry about the environment variable being the only way in which we differentiate CoreClr vs. Desktop. I think it would be safer to have the check be on CoreClr + `%DOTNET_HOST_PATH%`. 

Believe there are existing APIs in our Repo that can provide whether or not this is CoreClr. @agocke will be able to point out the appropriate one.